deb/format.go

```diff
@@ -257,51 +347,84 @@ func NewControlFileReader(r io.Reader, isRelease, isInstaller bool) *ControlFile
 	}
 }
 
-// ReadStanza reeads one stanza from control file
+// ReadStanza reads one stanza from control file
 func (c *ControlFileReader) ReadStanza() (Stanza, error) {
-	stanza := make(Stanza, 32)
+	buf := make(Stanza, 32)
+	err := c.ReadBufferedStanza(buf)
+	if err != nil {
+		return nil, err
+	}
+	if !buf.Empty() {
+		return buf, nil
+	}
+	return nil, nil
+}
+
+// ReadBufferedStanza reads one stanza from control file into the provided stanza
+func (c *ControlFileReader) ReadBufferedStanza(stanza Stanza) error {
+	if stanza == nil {
+		return ErrInvalidArgument
+	}
+
 	lastField := ""
 	lastFieldMultiline := c.isInstaller
 
 	for c.scanner.Scan() {
-		line := c.scanner.Text()
+		lineBytes := c.scanner.Bytes()
 
 		// Current stanza ends with empty line
-		if line == "" {
-			if len(stanza) > 0 {
-				return stanza, nil
+		if len(lineBytes) == 0 {
+			if !stanza.Empty() {
+				return nil
 			}
 			continue
 		}
 
-		if line[0] == ' ' || line[0] == '\t' || c.isInstaller {
+		if lineBytes[0] == ' ' || lineBytes[0] == '\t' || c.isInstaller {
+			if stanza[lastField] == nil {
+				stanza[lastField] = &strings.Builder{}
+			}
+
+			stanza[lastField].Grow(len(lineBytes) + 1)
 			if lastFieldMultiline {
-				stanza[lastField] += line + "\n"
+				stanza[lastField].Write(lineBytes)
+				stanza[lastField].WriteByte('\n')
 			} else {
-				stanza[lastField] += " " + strings.TrimSpace(line)
+				stanza[lastField].WriteByte(' ')
+				stanza[lastField].Write(bytes.TrimSpace(lineBytes))
 			}
 		} else {
-			parts := strings.SplitN(line, ":", 2)
-			if len(parts) != 2 {
-				return nil, ErrMalformedStanza
+			lineStr := *(*string)(unsafe.Pointer(&lineBytes))
+			splitIndex := strings.IndexByte(lineStr, ':')
+			if splitIndex == -1 {
+				stanza = nil
+				return ErrMalformedStanza
 			}
-			lastField = canonicalCase(parts[0])
+
+			// It's safe to pass a pointer to the lastField's underlying byte array
+			// to canonicalCase because canonicalCase is guaranteed to return a new string
+			lastFieldBytes := lineBytes[:splitIndex]
+			lastField = canonicalCase(*(*string)(unsafe.Pointer(&lastFieldBytes)))
 			lastFieldMultiline = isMultilineField(lastField, c.isRelease)
+
+			if stanza[lastField] == nil {
+				stanza[lastField] = &strings.Builder{}
+			}
+
+			lastFieldValue := lineBytes[splitIndex+1:]
 			if lastFieldMultiline {
-				stanza[lastField] = parts[1]
-				if parts[1] != "" {
-					stanza[lastField] += "\n"
+				stanza[lastField].Reset()
+				stanza[lastField].Grow(len(lastFieldValue) + 1)
+				stanza[lastField].Write(lastFieldValue)
+				if len(lastFieldValue) > 0 {
+					stanza[lastField].WriteByte('\n')
 				}
 			} else {
-				stanza[lastField] = strings.TrimSpace(parts[1])
+				stanza[lastField].Reset()
+				stanza[lastField].Grow(len(lastFieldValue))
```

Comment:
It might not save much, but we could do `bytes.TrimSpace` first, and then use the `len` of that to precisely allocate here.